### ISSUE TITLE
Add list-analagous constructors for vectors and hash tables

### ISSUE THREAD

**Issue by ornamentist** [coalton-lang/coalton#539]
Since Coalton provides the `make-list` list constructor, could we also have analogous `make-vector` and `make-hash-table` constructors? 

I understand the `make-hash-table` one is probably more subtle than it sounds, needing a convention over order of keys and values. Are there other Lisps conventions we can draw on?

**Comment by eliaslfox** [comment]
I've been thinking about this. I would like to implement a system like Haskell's [overloaded lists](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/overloaded_lists.html). Then we could add instances like

```
(define-class (FromList :a :b)
  (fromList (List :a -> :b :a))
  (fromListN (UFix -> List :a -> :b :a)))

(define-instance (FromList :a (List :a)))
(define-instance (FromList :a (Vector :a)))
(define-instance (Hash :a => (FromList (Tuple :a :b) (HashMap :a :b))))
```

For syntax I want to use reader macros. Probably `#[1 2 3 4]` so that `#(1 "hello" None)` can be a three element tuple.

**Comment by ornamentist** [comment]
That would be excellent.

**Comment by ornamentist** [comment]
Are there any caveats or issues with creating reader macros for Coalton code?  I can see some other (library-specific) uses for them but won't go down that path if they're too problematic.

**Comment by stylewarning** [comment]
> Are there any caveats or issues with creating reader macros for Coalton code?  I can see some other (library-specific) uses for them but won't go down that path if they're too problematic.

I don't think there's a problem with reader macros on their own, but there's a problem if Coalton users are forced to use them.

**Comment by gefjon** [comment]
Personally, I'm very wary of using reader macros for things like this -- they have a lot of weird subtle interactions with other parts of the ecosystem, including pretty bad editor and debugger support. Some concerns I have:

- Does `#[a b c]` construct a new list/vector/etc at read-time or at runtime? If read-time, you get subtle bugs where types aren't defined yet, and you violate non-lispers' expectations about what happens when you write `(let ((a b)) #[a b])`. If runtime, you violate lispers' expectations about how reader macros work, and I don't see how you implement the construction efficiently (i.e. without consing a list as an intermediate sequence to pass to `make-array` or friends).
- I found the prevalence of non-alphanumeric operators in Haskell a huge barrier to learning the language -- I had to look at documentation way more frequently because I didn't yet remember what some of the obscure glyphs did, and I always wished they were just replaced with a word. Is `#[1 2 3]` actually more readable and ergonomic than `(vec:make-vector 1 2 3)`, or is it just a neat feature that we could show off to other programming language people?
- Are hash-tables actually similar enough to flat sequences (lists and vectors) to use the same interface? I think the pitch that would convince me that overloaded list literals are a good idea is actual polymorphism: being able to write something like `(declare small-primes ((FromList :seq) => (:seq UFix))) (define small-primes #[2 3 5 7])`, or `(declare tuple-to-seq ((FromList :seq) => ((Tuple :elt :elt) -> (:seq :elt))) (define (tuple-to-seq tpl) (match tpl ((Tuple a b) #[a b])))`. It seems to me like hash-tables are meaningfully different from flat sequences, and trying to use the same interface to construct them is likely to lead to sadness. E.g. I think having the same key statically appear twice in a hash-table literal should be an error at compile-time, likely signaled by the `make-hash-table` macro.
- `(the (HashTable Foo Bar) #[#(foo bar) #(baz quux) #(a b) #(c d)])` is super ugly, and my Emacs/Sly setup doesn't indent it properly if I put each tuple on a new line. `(ht:make-hash-table (foo bar) (baz quux) (a b) (c d))` has far fewer superfluous and scary-looking sharp signs.

> Are there any caveats or issues with creating reader macros for Coalton code? I can see some other (library-specific) uses for them but won't go down that path if they're too problematic.

I don't think reader macros are any *more* problematic in Coalton code than in Common Lisp, but I think they're pretty problematic in Common Lisp...  I've been writing CL code for almost half of my life at this point, and I've reached a place where my instant reaction whenever I see a reader-dispatch sequence other than `#'`, `#+`, `#-`, `#.`, `#:`, `#x`, `#b` or `#\` is instant distress and distrust of whatever code I'm looking at.

**Comment by ornamentist** [comment]
> > Are there any caveats or issues with creating reader macros for Coalton code?  I can see some other (library-specific) uses for them but won't go down that path if they're too problematic.
> 
> I don't think there's a problem with reader macros on their own, but there's a problem if Coalton users are forced to use them.

Good point. I was thinking about reader macro synonyms for existing functions like `make-list` or `make-vector` which would be optional for users of the library.

**Comment by ornamentist** [comment]
> Personally, I'm very wary of using reader macros for things like this -- they have a lot of weird subtle interactions with other parts of the ecosystem, including pretty bad editor and debugger support. Some concerns I have:

Thank you for taking the time to record these points, they're very useful.

>     * Does `#[a b c]` construct a new list/vector/etc at read-time or at runtime? If read-time, you get subtle bugs where types aren't defined yet, and you violate non-lispers' expectations about what happens when you write `(let ((a b)) #[a b])`. If runtime, you violate lispers' expectations about how reader macros work, and I don't see how you implement the construction efficiently (i.e. without consing a list as an intermediate sequence to pass to `make-array` or friends).

Good point, I'd hoped we could have the reader macros for common data structures like Clojure has, but I understand Coalton and CL are a different environment.
 
>     * I found the prevalence of non-alphanumeric operators in Haskell a huge barrier to learning the language -- I had to look at documentation way more frequently because I didn't yet remember what some of the obscure glyphs did, and I always wished they were just replaced with a word. Is `#[1 2 3]` actually more readable and ergonomic than `(vec:make-vector 1 2 3)`, or is it just a neat feature that we could show off to other programming language people?

Agreed, I started going down the path of an "english.hs" Haskell module to replace the various glyphs with alphanumeric names. 
 
>     * Are hash-tables actually similar enough to flat sequences (lists and vectors) to use the same interface? I think the pitch that would convince me that overloaded list literals are a good idea is actual polymorphism: being able to write something like `(declare small-primes ((FromList :seq) => (:seq UFix))) (define small-primes #[2 3 5 7])`, or `(declare tuple-to-seq ((FromList :seq) => ((Tuple :elt :elt) -> (:seq :elt))) (define (tuple-to-seq tpl) (match tpl ((Tuple a b) #[a b])))`. It seems to me like hash-tables are meaningfully different from flat sequences, and trying to use the same interface to construct them is likely to lead to sadness. E.g. I think having the same key statically appear twice in a hash-table literal should be an error at compile-time, likely signaled by the `make-hash-table` macro.
> 
>     * `(the (HashTable Foo Bar) #[#(foo bar) #(baz quux) #(a b) #(c d)])` is super ugly, and my Emacs/Sly setup doesn't indent it properly if I put each tuple on a new line. `(ht:make-hash-table (foo bar) (baz quux) (a b) (c d))` has far fewer superfluous and scary-looking sharp signs.

Yes, hash tables are different enough from lists, vectors, sets etc they will need their own type-specific setup.

> 
> 
> > Are there any caveats or issues with creating reader macros for Coalton code? I can see some other (library-specific) uses for them but won't go down that path if they're too problematic.
> 
> I don't think reader macros are any _more_ problematic in Coalton code than in Common Lisp, but I think they're pretty problematic in Common Lisp... I've been writing CL code for almost half of my life at this point, and I've reached a place where my instant reaction whenever I see a reader-dispatch sequence other than `#'`, `#+`, `#-`, `#.`, `#:`, `#x`, `#b` or `#\` is instant distress and distrust of whatever code I'm looking at.

FWIW I found when writing test fixtures the amount of boilerplate made me start thinking about macros of some kind:

```
(define set-3 (the (Set Integer) (make-set (make-list 1 2 3))))
; ...
; tests using set-3
```

And wishing I could just:

```
; write test using #{ 1 2 3}
```

**Comment by gefjon** [comment]
> FWIW I found when writing test fixtures the amount of boilerplate made me start thinking about macros of some kind:

I think some ergonomic s-expression macros to improve testing would go a long way, and I'd love to have them. I just don't want reader macros.

I think your example code could be greatly simplified as:

```
(declare set-3 (Set Integer))
(define set-3 (make-set 1 2 3))
```

Where `make-set` is an s-expression macro analogous to `make-list`, or the `make-vector` and `make-hash-table` I drafted in #663. The `declare` is likely unnecessary; `make-set` plus usage should provide enough data to the type inference engine to make things work without it.

> `; write test using #{ 1 2 3}`

I really don't find `#{1 2 3}` better than `(make-set 1 2 3)` in any way, personally. With the `make-set` version, a reader can tell what's going on, and if they need more information, docstring lookup and jump-to-definition on `make-set` are just a key-chord away. With the `#{` version, you save typing a handful of letters (if that matters to you, configure your editor's autocomplete!), but a reader can no longer tell what sort of object is being constructed, whether there are any compile-time or runtime effects, whether you're doing something hacky with the type system, whether `1` `2` `3` are even being read as integers at all, etc, and figuring out the answers to those questions is a royal pain in the ass.

**Comment by ornamentist** [comment]
> > FWIW I found when writing test fixtures the amount of boilerplate made me start thinking about macros of some kind:
> 
> I think some ergonomic s-expression macros to improve testing would go a long way, and I'd love to have them. I just don't want reader macros.
> 
> I think your example code could be greatly simplified as:
> 
> ```
> (declare set-3 (Set Integer))
> (define set-3 (make-set 1 2 3))
> ```
> 
> Where `make-set` is an s-expression macro analogous to `make-list`, or the `make-vector` and `make-hash-table` I drafted in #663. The `declare` is likely unnecessary; `make-set` plus usage should provide enough data to the type inference engine to make things work without it.

I think this is the vector macro from that PR?

```
(cl:defmacro make-vector (cl:&rest elements)
  (cl:let* ((length (cl:length elements))
            (vec (cl:gensym "VEC-")))
    `(progn
      (let ,vec = (with-capacity ,length))
      ,@(cl:loop :for elt :in elements
           :collect `(push! ,elt ,vec))
      ,vec)))
```

**Comment by ornamentist** [comment]
> > FWIW I found when writing test fixtures the amount of boilerplate made me start thinking about macros of some kind:
> 
> I think some ergonomic s-expression macros to improve testing would go a long way, and I'd love to have them. I just don't want reader macros.
> 
> I think your example code could be greatly simplified as:
> 
> ```
> (declare set-3 (Set Integer))
> (define set-3 (make-set 1 2 3))
> ```
> 
> Where `make-set` is an s-expression macro analogous to `make-list`, or the `make-vector` and `make-hash-table` I drafted in #663. The `declare` is likely unnecessary; `make-set` plus usage should provide enough data to the type inference engine to make things work without it.

The other approach that would help with this kind of testing is some kind of Quickcheck style generative testing framework, but that's a whole big project just on its own.

**Comment by eliaslfox** [comment]
Fixed in #663
